source/slang/slang-ir-layout.cpp

```diff
@@ -193,7 +193,7 @@ case kIROp_##TYPE##Type:                                        \
                 offset = rules->adjustOffsetForNextAggregateMember(offset, fieldTypeLayout.alignment);
             }
         }
-        *outSizeAndAlignment = rules->alignCompositeElement(structLayout);
+        *outSizeAndAlignment = rules->alignCompositeElementOfAggregate(structLayout);
```

Comment:
I don't quite understand the need for alignCompositeElementOfAggregate here. Why can't we just rely on adjustOffsetForNextAggregateMember?

Reply:
`ConstantBuffer`s requires an alignment of 16 (cannot use partial registers) on all structs, arrays & matrixes (treated like an array). This functionality is equal to how `alignCompositeElement(resultLayout);` functioned and was intended to be used. The issue that we had was that `alignCompositeElement` considers `vector` as the same as `struct` for alignment rules (ConstantBuffer treats these types differently).

Reply:
Does that mean the current d3d constant buffer layout implementation is incorrect?

Reply:
Isn't the difference of vector types handled by getVectorSizeAndAlignment?

Reply:
> Isn't the difference of vector types handled by getVectorSizeAndAlignment?

How the old code worked was: `getVectorSizeAndAlignment` gets the vector size (and alignment), `alignCompositeElement` did alignment logic (since std140/std430 computes 'vector size' and 'alignment' differently)

Reply:
> Does that mean the current d3d constant buffer layout implementation is incorrect?

I believe our current implementation does not work as intended (array, matrix, and struct).

Example:
Given the following struct (you may copy paste this struct into https://maraneshi.github.io/HLSL-ConstantBufferLayoutVisualizer/):
```
cbuffer Test
{
    uint v0;
    float3 v1;
    uint3 v2;
    uint2 v3;
    uint2 v4;
    uint v5[4];
    uint3 v6[2];
    uint v7;
    uint2 v8;
    uint v9[2];
};
```

This struct currently (without this PR with `-fvk-use-dx-layout`) will compile into:
```
OpMemberName %SLANG_ParameterGroup_Test_std140 0 "v0"
OpMemberName %SLANG_ParameterGroup_Test_std140 1 "v1"
OpMemberName %SLANG_ParameterGroup_Test_std140 2 "v2"
OpMemberName %SLANG_ParameterGroup_Test_std140 3 "v3"
OpMemberName %SLANG_ParameterGroup_Test_std140 4 "v4"
OpMemberName %SLANG_ParameterGroup_Test_std140 5 "v5"
OpMemberName %SLANG_ParameterGroup_Test_std140 6 "v6"
OpMemberName %SLANG_ParameterGroup_Test_std140 7 "v7"
OpMemberName %SLANG_ParameterGroup_Test_std140 8 "v8"
OpMemberName %SLANG_ParameterGroup_Test_std140 9 "v9"

OpMemberDecorate %SLANG_ParameterGroup_Test_std140 0 Offset 0
OpMemberDecorate %SLANG_ParameterGroup_Test_std140 1 Offset 16
OpMemberDecorate %SLANG_ParameterGroup_Test_std140 2 Offset 32
OpMemberDecorate %SLANG_ParameterGroup_Test_std140 3 Offset 48
OpMemberDecorate %SLANG_ParameterGroup_Test_std140 4 Offset 56
OpMemberDecorate %SLANG_ParameterGroup_Test_std140 5 Offset 64
OpMemberDecorate %SLANG_ParameterGroup_Test_std140 6 Offset 128
OpMemberDecorate %SLANG_ParameterGroup_Test_std140 7 Offset 160
OpMemberDecorate %SLANG_ParameterGroup_Test_std140 8 Offset 168
OpMemberDecorate %SLANG_ParameterGroup_Test_std140 9 Offset 176
```
Which from what I can tell is wrong as per https://maraneshi.github.io/HLSL-ConstantBufferLayoutVisualizer/ and Shader-Playground.

With this PR and `fvk-use-dx-layout` the SPIR-V offsets are now correctly:
```
OpMemberName %SLANG_ParameterGroup_Test_std140 0 "v0"
OpMemberName %SLANG_ParameterGroup_Test_std140 1 "v1"
OpMemberName %SLANG_ParameterGroup_Test_std140 2 "v2"
OpMemberName %SLANG_ParameterGroup_Test_std140 3 "v3"
OpMemberName %SLANG_ParameterGroup_Test_std140 4 "v4"
OpMemberName %SLANG_ParameterGroup_Test_std140 5 "v5"
OpMemberName %SLANG_ParameterGroup_Test_std140 6 "v6"
OpMemberName %SLANG_ParameterGroup_Test_std140 7 "v7"
OpMemberName %SLANG_ParameterGroup_Test_std140 8 "v8"
OpMemberName %SLANG_ParameterGroup_Test_std140 9 "v9"

OpMemberDecorate %SLANG_ParameterGroup_Test_default 0 Offset 0
OpMemberDecorate %SLANG_ParameterGroup_Test_default 1 Offset 4
OpMemberDecorate %SLANG_ParameterGroup_Test_default 2 Offset 16
OpMemberDecorate %SLANG_ParameterGroup_Test_default 3 Offset 32
OpMemberDecorate %SLANG_ParameterGroup_Test_default 4 Offset 40
OpMemberDecorate %SLANG_ParameterGroup_Test_default 5 Offset 48
OpMemberDecorate %SLANG_ParameterGroup_Test_default 6 Offset 112
OpMemberDecorate %SLANG_ParameterGroup_Test_default 7 Offset 140
OpMemberDecorate %SLANG_ParameterGroup_Test_default 8 Offset 144
OpMemberDecorate %SLANG_ParameterGroup_Test_default 9 Offset 160
```

Reply:
Well, you are running spirv compiles which will never trigger the existing HLSL constant buffer layout logic. You will need to use reflection test and compile it to d3d target to see if the layout computed by slang is correct. In that case the existing HLSL constant buffer layout code will be executed. If you just compile to spirv you are just testing the std140 logic, as indicated in the spirv you pasted.

Reply:
HLSL has correct reflection data.

Reply:
Reflection is based on AST, AST has correct data.

What happened was that the SizeAndAlignment setup (IR-pass) was producing the wrong values for HLSL. Since we don't care about size/alignment for HLSL emitting this was not an issue